### PR TITLE
Jacob branch new 

### PR DESCRIPTION
**Pull request by quickTwix898**
Added tool for kuri to speak, either from preloaded phrases or online from Amazon Polly. Online and offline modes. New libraries added; NuGet and AWSSDKs.

Added a new scene with KuriTextManager, called JacobMain.

Restored KuriTextManager

Added kuri speech components to scripts folder

Added preload mp3 files

Added caching aws downloads feature

Cleaned up and deleted classes for PR

### CODE REVIEW

File: Assets/NuGet/LICENSE [hunk 1]
```diff
@@ -0,0 +1,21 @@
+MIT License
```
Comment (tgroechel) [inline Assets/NuGet/LICENSE:1]: Could you update the readme regarding the licenses?

File: Assets/Scripts/AWSPollyGetter.cs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+using UnityEngine;
+using Amazon.Polly;
+using Amazon.Polly.Model;
+using System.IO;
+using System;
+using System.Text;
+
+namespace MoveToCode {
+    public class AWSPollyGetter : Singleton<AWSPollyGetter> {
+        private const string APIPath = "Assets/Resources/PollyAPI1.txt";
```
Comment (tgroechel) [inline Assets/Scripts/AWSPollyGetter.cs:12]: Move this to `ResourcePathConstants.cs`

File: Assets/Scripts/AWSPollyGetter.cs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+        private const string speakstyle = "pitch=\"x-high\" rate=\"100%\"";
+        private VoiceId speakvoice = VoiceId.Joanna;
+
+        private AmazonPollyClient apc;
+        public bool isWorking = true;
+        private bool onlineState;
+
+        public AWSPollyGetter(bool onlineState) {
+            this.onlineState = onlineState;
```
Comment (tgroechel) [inline Assets/Scripts/AWSPollyGetter.cs:22]: get rid of all instances of "this.", it is implied so it is actually bad form

File: Assets/Scripts/AWSPollyGetter.cs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        private const string APIPath = "Assets/Resources/PollyAPI1.txt";
+
+        private const string speakstyle = "pitch=\"x-high\" rate=\"100%\"";
+        private VoiceId speakvoice = VoiceId.Joanna;
+
+        private AmazonPollyClient apc;
+        public bool isWorking = true;
+        private bool onlineState;
+
+        public AWSPollyGetter(bool onlineState) {
```
Comment (tgroechel) [inline Assets/Scripts/AWSPollyGetter.cs:21]: rename parameter to be onlineStateIn

File: Assets/Scripts/AWSPollyGetter.cs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+namespace MoveToCode {
+    public class AWSPollyGetter : Singleton<AWSPollyGetter> {
+        private const string APIPath = "Assets/Resources/PollyAPI1.txt";
+
+        private const string speakstyle = "pitch=\"x-high\" rate=\"100%\"";
+        private VoiceId speakvoice = VoiceId.Joanna;
+
+        private AmazonPollyClient apc;
+        public bool isWorking = true;
```
Comment (tgroechel) [inline Assets/Scripts/AWSPollyGetter.cs:18]: rename to what is working exactly

File: Assets/Scripts/AWSPollyGetter.cs (hunk 2 above)
Comment (tgroechel) [reply Assets/Scripts/AWSPollyGetter.cs:12]: also call it `AWSPollyAPI`

File: Assets/Scripts/AWSPollyGetter.cs [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        private bool onlineState;
+
+        public AWSPollyGetter(bool onlineState) {
+            this.onlineState = onlineState;
+            if (onlineState) { 
+                StreamReader reader = new StreamReader(APIPath);
+                try {
+                    apc = new AmazonPollyClient(reader.ReadLine(), reader.ReadLine(), Amazon.RegionEndpoint.USWest1);
```
Comment (tgroechel) [inline Assets/Scripts/AWSPollyGetter.cs:26]: can you move the region out to the config file with the aws config (aka where you put your api keys)

File: Assets/Scripts/AWSPollyGetter.cs [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+            if (isWorking) {
+                SynthesizeSpeechRequest sreq = new SynthesizeSpeechRequest() {
+                    TextType = TextType.Ssml,
+                    Text = "<speak><prosody " + speakstyle + ">" + lyric + "</prosody></speak>",
+                    VoiceId = speakvoice,
+                    OutputFormat = OutputFormat.Mp3
+                };
+
+                StringBuilder sb = new StringBuilder();
+                foreach (char c in lyric) { //TODO stop the spaghetti
```
Comment (tgroechel) [inline Assets/Scripts/AWSPollyGetter.cs:49]: yeah fix the spaghetti before the PR (you should resolve most TODO comments before PR in general)

File: Assets/Scripts/AWSPollyGetter.cs [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+                return ResourcePathConstants.SpeechCacheFolder + sb.ToString();
+            } else {
+                throw new FileLoadException();
+            }
+        }
+
+        private void OnlineRequestPolly(SynthesizeSpeechRequest sreq, string name) {
+            SynthesizeSpeechResponse sres = apc.SynthesizeSpeech(sreq);
+            using (FileStream fileStream = File.Create(@"Assets\Resources\Audio\CacheSpeech\" + name + ".mp3")) {
```
Comment (tgroechel) [inline Assets/Scripts/AWSPollyGetter.cs:68]: move the path into `PathResourceConstants.cs`

File: Assets/Scripts/KuriPhrase.cs [hunk 9]
```diff
@@ -0,0 +1,32 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+
```
Comment (tgroechel) [inline Assets/Scripts/KuriPhrase.cs:5]: remove excess line

File: Assets/Scripts/KuriPhrase.cs [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+        public string filepath;
+
+        public void Play(AudioSource auds) { // use audiosource bundled with KuriPhraseManager 
+            AudioClip ac = Resources.Load<AudioClip>(filepath);
+            auds.clip = ac;
+            auds.Play();
+        }
+
+        public KuriPhrase(int id, string lyric, USECASE uc, string filepath) {
+            this.id = id;
```
Comment (tgroechel) [inline Assets/Scripts/KuriPhrase.cs:26]: remove `this.`

File: Assets/Scripts/KuriPhraseManager.cs [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+using UnityEditor;
+using UnityEngine;
+using Newtonsoft.Json;
+
+namespace MoveToCode {
+    public class KuriPhraseManager : Singleton<KuriPhraseManager> {
+        private List<KuriPhrase> phrases; 
+        private List<KuriPhrase> preloadEncouragement;
+        private List<KuriPhrase> preloadCongratulation;
+        private AWSPollyGetter apg;
```
Comment (tgroechel) [inline Assets/Scripts/KuriPhraseManager.cs:13]: spell this out to `awsPollyGetter` as opposed to `apg`

File: Assets/Scripts/KuriPhraseManager.cs [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+        private KuriPhrase empty = new KuriPhrase(-1, "", KuriPhrase.USECASE.Other, "emptyfile");
+        
+        public KuriPhrase GetPhrase(string lyric) {
+            foreach (KuriPhrase kp in phrases) {
+                if(kp.lyric.Equals(lyric)) {
+                    return kp;
+                }
+            }
+            //if phrase does not exist yet
+            if(apg.isWorking) {
```
Comment (tgroechel) [inline Assets/Scripts/KuriPhraseManager.cs:29]: You made a getter for `isWorking`, use that

File: Assets/Scripts/KuriPhrase.cs [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+            Congratulation,
+            Other
+        };
+
+        public int id;
+        public string lyric;
+        public USECASE uc;
+        public string filepath;
+
+        public void Play(AudioSource auds) { // use audiosource bundled with KuriPhraseManager
```
Comment (tgroechel) [inline Assets/Scripts/KuriPhrase.cs:19]: rename function so that the comment is not necessary

File: Assets/Scripts/KuriPhraseManager.cs [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+using System.IO;
+using UnityEditor;
+using UnityEngine;
+using Newtonsoft.Json;
+
+namespace MoveToCode {
+    public class KuriPhraseManager : Singleton<KuriPhraseManager> {
+        private List<KuriPhrase> phrases; 
+        private List<KuriPhrase> preloadEncouragement;
+        private List<KuriPhrase> preloadCongratulation;
```
Comment (tgroechel) [inline Assets/Scripts/KuriPhraseManager.cs:12]: make these `XList` eg `preloadCongfigurationList`

File: Assets/Scripts/KuriPhraseManager.cs [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+        private AWSPollyGetter apg;
+        private int idCount;
+        public bool onlineMode = true;
+
+
```
Comment (tgroechel) [inline Assets/Scripts/KuriPhraseManager.cs:17]: remove space

File: Assets/Scripts/KuriPhraseManager.cs [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+        [HideInInspector] public AudioSource audioSource;
```
Comment (tgroechel) [inline Assets/Scripts/KuriPhraseManager.cs:18]: why make it public if you are hiding?

File: Assets/Scripts/AWSPollyGetter.cs (hunk 5 above)
Comment (tgroechel) [reply Assets/Scripts/AWSPollyGetter.cs:18]: also make this private varriable

File: Assets/Scripts/KuriPhraseManager.cs [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+
+        private KuriPhrase empty = new KuriPhrase(-1, "", KuriPhrase.USECASE.Other, "emptyfile");
+        
+        public KuriPhrase GetPhrase(string lyric) {
+            foreach (KuriPhrase kp in phrases) {
+                if(kp.lyric.Equals(lyric)) {
+                    return kp;
+                }
+            }
+            //if phrase does not exist yet
```
Comment (tgroechel) [inline Assets/Scripts/KuriPhraseManager.cs:28]: shouldn't need this comment if you refactor the name/getter function for `isWorking`

File: Assets/Scripts/KuriPhraseManager.cs [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+
+                using (StreamWriter sw = new StreamWriter(
+                    File.Open("Assets/Resources/" + ResourcePathConstants.SpeechCacheFolder + "cachePhrase.json", FileMode.Truncate))) {
+                    sw.WriteLine(JsonConvert.SerializeObject(phrases, Formatting.None));
+                    sw.Close();
+                };
+
+                return newPhrase;
+            } else {
+                KuriTextManager ktm = FindObjectOfType<KuriTextManager>();
```
Comment (tgroechel) [inline Assets/Scripts/KuriPhraseManager.cs:51]: you should avoid `FindObjectOfType`, the best part of singleton is you can actually just reference it via `KuriTextManager.instance`

File: Assets/Scripts/KuriPhraseManager.cs [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+                ktm.Addline(lyric);
+                return empty; 
+            }
+        }
+
+        public KuriPhrase GetPhrase(KuriPhrase.USECASE ucase) {
+            if(ucase != KuriPhrase.USECASE.Other) {
+                if(ucase == KuriPhrase.USECASE.Congratulation) {
+                    int temp = (int)(Random.value * preloadCongratulation.Count);
+                    Debug.Log(temp);
```
Comment (tgroechel) [inline Assets/Scripts/KuriPhraseManager.cs:61]: remove all debug statements

Comment (tgroechel) [reply Assets/Scripts/KuriPhraseManager.cs:61]: unless you feel they are absolutely needed/would help people debug via LogWarning/Error

File: Assets/Scripts/KuriPhraseManager.cs [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+                    return preloadCongratulation[temp];
+                    
+                } else if(ucase == KuriPhrase.USECASE.Encouragement) {
```
Comment (tgroechel) [inline Assets/Scripts/KuriPhraseManager.cs:64]: Please move `else if` you next line eg:

Comment (tgroechel) [reply Assets/Scripts/KuriPhraseManager.cs:64]:
```
if(2 > x){
}
else{
}
```

File: Assets/Scripts/KuriPhraseManager.cs [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+                if(kp.lyric.Equals(lyric)) {
+                    return kp;
+                }
+            }
+            //if phrase does not exist yet
+            if(apg.isWorking) {
+                string filepath;
+                try {
+                    filepath = apg.PullPhrase(lyric);
+                } catch(FileLoadException e) {
```
Comment (tgroechel) [inline Assets/Scripts/KuriPhraseManager.cs:33]: move catch to next line

File: Assets/Scripts/AWSPollyGetter.cs [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+                    }
+                }
+                try {
+                    OnlineRequestPolly(sreq, sb.ToString());
+                } catch(Exception e) {
+                    throw new FileLoadException();
+                }
+
+                return ResourcePathConstants.SpeechCacheFolder + sb.ToString();
+            } else {
```
Comment (tgroechel) [inline Assets/Scripts/AWSPollyGetter.cs:61]: else on next line

File: Assets/Scripts/KuriPhraseManager.cs (hunk 20 above)
Comment (tgroechel) [reply Assets/Scripts/KuriPhraseManager.cs:64]: please do for rest as well

File: Assets/Scripts/KuriPhraseManager.cs [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+                } else {
+                    throw new System.Exception("You have some problem with the KuriPhrase.USECASE enum");
+                }
+            } else {
+                Debug.LogWarning("Warning: \"other\" usecase phrase was specifically requested");
+                return empty;
+            }
+        }
+
+        void Awake() {
```
Comment (tgroechel) [inline Assets/Scripts/KuriPhraseManager.cs:75]: awake should always be the first function, after member variables

File: Assets/Scripts/KuriPhraseManager.cs [hunk 24]
```diff
@@ -0,0 +1,102 @@
...
+            if (onlineMode) {
+                apg = new AWSPollyGetter(onlineMode);
+            }
+            else {
+                apg = new AWSPollyGetter(onlineState: false);
+            }
+            idCount = 0;
+            phrases = new List<KuriPhrase>(); 
+            StreamReader reader = new StreamReader("Assets/Resources/" + ResourcePathConstants.SpeechFolder + "preloadEncouragement.json"); 
+            //TODO filepaths that work post-build
```
Comment (tgroechel) [inline Assets/Scripts/KuriPhraseManager.cs:85]: careful leaving in comments inline like this unless needed

File: Assets/Scripts/KuriPhraseManager.cs [hunk 25]
```diff
@@ -0,0 +1,102 @@
...
+            else {
+                apg = new AWSPollyGetter(onlineState: false);
+            }
+            idCount = 0;
+            phrases = new List<KuriPhrase>(); 
+            StreamReader reader = new StreamReader("Assets/Resources/" + ResourcePathConstants.SpeechFolder + "preloadEncouragement.json"); 
+            //TODO filepaths that work post-build 
+            preloadEncouragement = JsonConvert.DeserializeObject<List<KuriPhrase>>(reader.ReadToEnd());
+
+            reader = new StreamReader("Assets/Resources/" + ResourcePathConstants.SpeechFolder + "preloadCongratulation.json");
```
Comment (tgroechel) [inline Assets/Scripts/KuriPhraseManager.cs:88]: move to path resource constants

File: Assets/Scripts/KuriPhraseManager.cs [hunk 26]
```diff
@@ -0,0 +1,102 @@
...
+            preloadCongratulation = JsonConvert.DeserializeObject<List<KuriPhrase>>(reader.ReadToEnd());
+
+            try {
+                reader = new StreamReader("Assets/Resources/" + ResourcePathConstants.SpeechCacheFolder + "cachePhrase.json");
+                phrases = JsonConvert.DeserializeObject<List<KuriPhrase>>(reader.ReadToEnd());
+            } catch(FileNotFoundException e) { // no cached phrases
```
Comment (tgroechel) [inline Assets/Scripts/KuriPhraseManager.cs:94]: remove the comment and make more explicit via code

File: Assets/Scripts/AWSPollyGetter.cs (hunk 6 above)
Comment (quickTwix898) [reply Assets/Scripts/AWSPollyGetter.cs:26]: it's a text file, so i'll have to make a dictionary

**Comment by quickTwix898** [comment]
Resolved mentioned issues

**Pull request merged**